File: problem_023.py
#!/usr/bin/env python

print("Project Euler - Problem 23:")
print("Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.\n")

limit = 28123

divisorSums = [0] * limit
abundantNumbers = []

# Use seive to find divisor sums
for i in range(1, limit):
    if divisorSums[i] > i:
        abundantNumbers.append(i)
    for n in range(i, limit, i):
        divisorSums[n] += i

# Combine abundant numbers and eliminate abundant sums
nonAbundantSums = [i for i in range(limit)]
for a in range(len(abundantNumbers)):
    for b in range(a, len(abundantNumbers)):
        abundantSum = abundantNumbers[a] + abundantNumbers[b]
        if abundantSum >= limit:
            break

        nonAbundantSums[abundantSum] = 0

print(sum(nonAbundantSums))

File: problem_023.lua
#!/usr/bin/env lua

print("Project Euler - Problem 23")
print("Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.\n")

function divSum(n)
   sum = 0
   for d = 1, n/2+1, 1 do
      if (n % d == 0) then
         sum = sum + d
      end
   end
   return sum
end

limit = 28123

abundant = {}
index = 0
for n = 12, limit, 1 do
   if divSum(n) > n then
      abundant[index] = n
      index = index + 1
   end
end

sums = {}
for a = 0, index-1, 1 do
   for b = a, index-1, 1 do
      n = abundant[a] + abundant[b]
      if n > limit then
         break
      end
      sums[n] = 1
   end
end

sum = 0
for i = 0, limit, 1 do
   if sums[i] == nil then
      sum = sum + i
   end
end

print(sum)
